### ISSUE TITLE
promote causes segfault or StackOverflow with bad promote_rules

### ISSUE THREAD

**Issue by omus** [JuliaLang/julia#19968]
I was messing around with a toy `promote_rule` example and I discovered a segfault on Julia 0.5 and `StackOverflow` in Julia 0.6
```julia
julia> type ModInt{n} <: Integer
               k::Int
               ModInt(k::Integer) = new(mod(k,n))
       end

julia> Base.promote_rule{n}(::Type{ModInt{n}}, ::Type{Int}) = ModInt{n}

julia> Base.promote_rule{n}(::Type{Int}, ::Type{ModInt{n}}) = Int
```
Julia 0.5
```julia
julia> promote(ModInt{4}(6), 7)
Segmentation fault: 11
```
Julia 0.6

```julia
julia> promote(ModInt{4}(6), 7)
ERROR: StackOverflowError:
Stacktrace:
 [1] promote_type(::Type{ModInt{4}}, ::Type{Int64}) at ./promotion.jl:149 (repeats 80000 times)
```

**Comment by tkelman** [comment]
Is your 0.6 build from before or after https://github.com/JuliaLang/julia/pull/19918?

**Comment by omus** [comment]
Looks to be before. I'll rebuild and report back.

**Comment by omus** [comment]
Still occurs after #19918 was merged

**Comment by StefanKarpinski** [comment]
It's pretty hard to generically catch this kind of circularity in dispatch. I'm not sure that we can do better than throwing a stack overflow for this.

**Comment by JeffBezanson** [comment]
#19918 added a circularity check to a new version of `promote`, but we could instead add it to the usual `promote` function. @timholy was being wisely cautious there.

Of course, I'm also happy to get a stack overflow instead of a segfault.

**Comment by JeffBezanson** [comment]
I should add that in 0.5, the problem is almost certainly also a stack overflow, but turning those into exceptions can be unreliable.

**Comment by omus** [comment]
A "won't fix" is a perfectly reasonable answer here. Ideally there would be a more helpful exception but I think the stack trace is good enough.

@JeffBezanson should this remain open as an issue for 0.5?

**Comment by KristofferC** [comment]
Seems that StackOverflow is ok here.
